src/mappings/horizonStaking.ts

```diff
@@ -279,4 +279,184 @@ export function handleTokensToDelegationPoolAdded(event: TokensToDelegationPoolA
+    if (provision.delegatorShares != BigInt.fromI32(0)) {
+        provision = updateDelegationExchangeRateForProvision(provision as Provision)
+    }
+    provision = updateAdvancedProvisionMetrics(provision as Provision)
+    provision.save()
+
+    // update indexer
+    let indexer = createOrLoadIndexer(event.params.serviceProvider, event.block.timestamp)
+    indexer.delegatedTokens = indexer.delegatedTokens.plus(event.params.tokens)
+    indexer.delegatorShares = indexer.delegatorShares.plus(event.params.shares)
+    indexer.save()
+
+    // update delegator
+    let delegatorID = event.params.delegator.toHexString()
+    let delegator = createOrLoadDelegator(event.params.delegator, event.block.timestamp)
+    delegator.totalStakedTokens = delegator.totalStakedTokens.plus(event.params.tokens)
+    delegator.save()
+
+    // update delegated stake
+    let delegatedStake = createOrLoadDelegatedStakeForProvision(
+        delegatorID,
+        indexer.id,
+        dataService.id,
+        event.block.timestamp.toI32(),
+    )
+
+    if (!zeroShares) {
+        let previousExchangeRate = delegatedStake.personalExchangeRate
+        let previousShares = delegatedStake.shareAmount
+        let averageCostBasisTokens = previousExchangeRate
+            .times(previousShares.toBigDecimal())
+            .plus(event.params.tokens.toBigDecimal())
+        let averageCostBasisShares = previousShares.plus(event.params.shares)
+        if (averageCostBasisShares.gt(BigInt.fromI32(0))) {
+            delegatedStake.personalExchangeRate = averageCostBasisTokens
+                .div(averageCostBasisShares.toBigDecimal())
+                .truncate(18)
+        }
+    }
+
+    let isStakeBecomingActive = delegatedStake.shareAmount.isZero() && !event.params.shares.isZero()
+
+    delegatedStake.stakedTokens = delegatedStake.stakedTokens.plus(event.params.tokens)
+    delegatedStake.shareAmount = delegatedStake.shareAmount.plus(event.params.shares)
+    delegatedStake.lastDelegatedAt = event.block.timestamp.toI32()
+    delegatedStake.save()
+
+    // reload delegator to avoid edge case where we can overwrite stakesCount if stake is new
+    delegator = Delegator.load(delegatorID) as Delegator
+
+    // upgrade graph network
+    let graphNetwork = createOrLoadGraphNetwork(event.block.number, event.address)
+    graphNetwork.totalDelegatedTokens = graphNetwork.totalDelegatedTokens.plus(event.params.tokens)
+
+    if (isStakeBecomingActive) {
+        graphNetwork.activeDelegationCount = graphNetwork.activeDelegationCount + 1
+        delegator.activeStakesCount = delegator.activeStakesCount + 1
+        // Is delegator becoming active because of the stake becoming active?
+        if (delegator.activeStakesCount == 1) {
+            graphNetwork.activeDelegatorCount = graphNetwork.activeDelegatorCount + 1
+        }
+    }
+
+    graphNetwork.save()
+    delegator.save()
+}
+
+export function handleDelegationSlashed(event: DelegationSlashed): void {
```

Comment:
We should probably add `tokensSlashedServiceProvider` and `tokensSlashedDelegationPool` to the `Provision` entity to track those amounts.